File: NEAT/node.py
from .gene import *
import math

class Node(object):
    def __init__(self, n, l, is_bias=False) -> None:
        self.number = n
        self.layer = l # type -> input, hidden bias

        self.output = 0
        self.is_bias = is_bias

        self.genes:list[Gene] = [] # type: ignore

        self.sigmoid = lambda x: 1 / (1 + math.exp(-x))

    def clone(self):
        n = Node(self.number, self.layer, self.is_bias)
        n.output = self.output
        return n

    def calculate(self):
        if self.layer == 0:
            return
        
        s = 0
        for g in self.genes:
            if g.enabled:
                s += g.in_node.output * g.weight

        self.output = self.sigmoid(s)

File: treejs_test/neat/node.ts
import { Gene } from "./gene.ts"

export class Node {
  output = 0
  activated = true
  genes: Gene[] = []

  constructor(
    public number: number,
    public layer: number
  ) { }

  private sigmoid(x:number) {
    return 1 / (1 + Math.exp(-x));
  }

  public clone() {
    const node = new Node(this.number, this.layer)
    node.output = this.output
    return node
  }

  public calculate() {
    if (this.layer == 0) {
      return
    }

    let s = 0
    this.genes.forEach(gene => {
      if (gene.enabled) s += gene.in_node.output * gene.weight
    });

    this.output = this.sigmoid(s)
  }
}